### PR TITLE
Add test run for openzfs

### PR DESCRIPTION
**Pull request by grisu48**
Adds a test for [filesystem/zfs](https://build.opensuse.org/package/show/filesystems/zfs) for openSUSE Leap

- Related ticket: https://progress.opensuse.org/issues/70903
- Needles: -
- Verification run: [Leap 15.2](http://phoenix-openqa.qam.suse.de/tests/2722)

Tumbleweed has a too recent Kernel for the ZoL modules: [Tumbleweed](http://phoenix-openqa.qam.suse.de/tests/2299#step/zfs/10)

### CODE REVIEW

File: tests/console/zfs.pm [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    assert_script_run('curl -v -o /tank/Big_Buck_Bunny_8_seconds_bird_clip.ogv ' . data_url('Big_Buck_Bunny_8_seconds_bird_clip.ogv'));
+    assert_script_run('md5sum Big_Buck_Bunny_8_seconds_bird_clip.ogv > /var/tmp/Big_Buck_Bunny_8_seconds_bird_clip.ogv.md5sum');
+    assert_script_run('md5sum -c /var/tmp/Big_Buck_Bunny_8_seconds_bird_clip.ogv.md5sum');
+    scrub('tank');
+    assert_script_run('zpool status tank | grep state | grep ONLINE');
+    script_run('zpool status tank');
+    save_screenshot();
+    ## Test raidz capability
+    corrupt_disk('/var/tmp/tank_a.img');
+    # Wait until degraded
```
Comment (pdostal) [inline tests/console/zfs.pm:94]: This does not make sense to me.

File: tests/console/zfs.pm [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    assert_script_run('md5sum -c /var/tmp/Big_Buck_Bunny_8_seconds_bird_clip.ogv.md5sum');
+    scrub('tank');
+    assert_script_run("zpool status tank | grep scan | grep 'scrub repaired'");
+    assert_script_run('md5sum -c /var/tmp/Big_Buck_Bunny_8_seconds_bird_clip.ogv.md5sum');
+    script_run('zpool status tank');
+    assert_script_run('zpool status tank | grep "corrupted data"');
+    # Replace disk
+    assert_script_run('zpool offline tank /var/tmp/tank_a.img');
+    assert_script_run('zpool replace tank /var/tmp/tank_a.img');
```
Comment (pdostal) [inline tests/console/zfs.pm:103]: Is it OK that you replaced one disk by the same so you actually didn't replaced it but just replugged it?

File: tests/console/zfs.pm [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+sub reboot {
+    my ($self) = @_;
+    power_action('reboot', textmode => 1);
+    $self->wait_boot(bootloader_time => 300);
+    select_console 'root-console';
+}
+
+sub run {
```
Comment (pdostal) [inline tests/console/zfs.pm:69]: I'd add some empty lines to separate each code block.

File: tests/console/zfs.pm [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    # Create second snapshot with other data
+    assert_script_run('zfs snapshot tank@second');
+    assert_script_run('zfs snapshot tank@third');
+    assert_script_run('zfs list -t snapshot | grep "tank@second"');
+    assert_script_run('zfs list -t snapshot | grep "tank@third"');
+    # Mount snapshots
+    assert_script_run('mkdir -p /mnt/tank/{initial,second}');
+    assert_script_run('mount -t zfs tank@initial /mnt/tank/initial');
+    assert_script_run('mount -t zfs tank@second /mnt/tank/second');
+    assert_script_run('stat /mnt/tank/initial/Big_Buck_Bunny_8_seconds_bird_clip.ogv');
```
Comment (pdostal) [inline tests/console/zfs.pm:123]: I'd try to do also `md5sum` to verify the file is still the same.

File: tests/console/zfs.pm [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    # Rename snapshot
+    assert_script_run('zfs rename tank@initial tank@today');
+    assert_script_run('zfs list -t snapshot | grep "tank@today"');
+    assert_script_run('! zfs list -t snapshot | grep "tank@initial"');
+    # Send snapshot to second zfs pool
+    assert_script_run('zfs send "tank@today" | zfs recv -F "dozer@today"');
+    script_run('zfs list -t snapshot | grep "dozer@today"');
+    ## Test rollback
+    assert_script_run('zfs rollback -r tank@today');
+    # Rollback to today also deletes more recent snaptshots
```
Comment (pdostal) [inline tests/console/zfs.pm:147]: I don't see any rollback here.

**Review by pdostal — CHANGES_REQUESTED** [review]

File: tests/console/zfs.pm (hunk 3 above)
Comment (grisu48) [reply tests/console/zfs.pm:69]: Good idea.

File: tests/console/zfs.pm (hunk 5 above)
Comment (grisu48) [reply tests/console/zfs.pm]: That comment belongs is a note to the applied rollback: `zfs rollback -r tank@today`

I updated the comment to make that better understandable

File: tests/console/zfs.pm (hunk 4 above)
Comment (grisu48) [reply tests/console/zfs.pm:123]: Good idea! I've updated that on all checks.

File: tests/console/zfs.pm (hunk 2 above)
Comment (grisu48) [reply tests/console/zfs.pm]: `zpool replace tank DISK` replaces the affected disk with a new disk in the same location, so it should be fine. It's similar to unplugging a hot-swappable disk and replace it with a new one.

See [Replacing a Device in a ZFS Storage Pool](https://docs.oracle.com/cd/E19253-01/819-5461/gbcet/index.html)

IMHO this should be fine.

File: tests/console/zfs.pm (hunk 1 above)
Comment (grisu48) [reply tests/console/zfs.pm]: Removed

File: tests/console/zfs.pm (hunk 2 above)
Comment (pdostal) [reply tests/console/zfs.pm]: No, you are replacing `/var/tmp/tank_a.img` by `/var/tmp/tank_a.img` or am I lost somewhere else? 😆

File: tests/console/zfs.pm [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    power_action('reboot', textmode => 1);
+    $self->wait_boot(bootloader_time => 300);
+    select_console 'root-console';
+}
+
+sub run {
+    # Preparation
+    my $self = shift;
+    # TODO: Switch to serial_terminal for BLAZING FAST EXECUTING!!!
+    #$self->select_serial_terminal;
```
[the remark this answers is not in the document]
Comment (pdostal) [reply tests/console/zfs.pm]: 👍

File: tests/console/zfs.pm [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    assert_script_run('! stat /tank/test_unzip.zip');
+    # Test snapshot after reboot
+    assert_script_run('zfs rollback -r tank@third');
+    assert_script_run('stat /tank/BBB_8_seconds_bird_clip.ogv');
+    assert_script_run('stat /tank/snapshot_3');
+}
+
+sub post_fail_hook {
+    cleanup();
+}
```
[the remark this answers is not in the document]
Comment (pdostal) [reply tests/console/zfs.pm]: 👍

File: tests/console/zfs.pm (hunk 2 above)
Comment (grisu48) [reply tests/console/zfs.pm]: `zpool replace /var/tmp/tank_a.img` **marks** this disk as being replaced and thus should be fine.

The procedure is the following:

    assert_script_run('zpool offline tank /var/tmp/tank_a.img');
    clear_disk('/var/tmp/tank_a.img');
    assert_script_run('zpool replace tank /var/tmp/tank_a.img');

This is the same as taking the disk offline, replace it and (somehow it ends up in the same location) then mark the disk as replaced. Think of `clear_disk` as "I take the disk out and hot-swap it with a new one". Think of `zpool replace` as "disk is marked as replaced"

Comment (pdostal) [reply tests/console/zfs.pm]: Ah yes, I missed the `clear_disk('/var/tmp/tank_a.img');` step. Sorry Felix!

**Review by pdostal — APPROVED** [review]

**Comment by grisu48** [comment]
I would like to leave this open until next week so the community has a chance to comment on this as well

File: tests/console/zfs.pm (hunk 2 above)
Comment (grisu48) [reply tests/console/zfs.pm]: You didn't miss it, I added it afterwards ;-)

**Comment by lkocman** [comment]
Fine by me if it doesn't block the release. Would be my fs of choice if there wouldn't be licensing issue.

**Comment by grisu48** [comment]
Agreed with @lkocman that this should be in the [Leap Development group](https://openqa.opensuse.org/group_overview/39) to ensure, this is not becoming a blocking issue for releases

File: tests/console/zfs.pm [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    assert_script_run('mkdir -p /mnt/tank/{initial,second}');
+    assert_script_run('mount -o ro -t zfs tank@initial /mnt/tank/initial');
+    assert_script_run('zfs clone -o mountpoint=/mnt/tank/second tank@second tank/2nd_second');
+    assert_script_run('stat /mnt/tank/initial/Big_Buck_Bunny_8_seconds_bird_clip.ogv');
+    assert_script_run('cd /mnt/tank/initial/; md5sum -c /var/tmp/Big_Buck_Bunny_8_seconds_bird_clip.ogv.md5sum');
+    assert_script_run('! stat /mnt/tank/initial/test_unzip.zip');
+    assert_script_run('! stat /mnt/tank/second/Big_Buck_Bunny_8_seconds_bird_clip.ogv');
+    assert_script_run('stat /mnt/tank/second/test_unzip.zip');
+    assert_script_run('cd');
+    assert_script_run('umount /mnt/tank/{initial,second}');
```
Comment (jankohoutek) [inline tests/console/zfs.pm:152]: clone named second should be unmout by zfs it self:
`zfs unmount tank/2nd_second`

File: tests/console/zfs.pm [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    assert_script_run('zfs list -t snapshot | grep "tank@second"');
+    assert_script_run('zfs list -t snapshot | grep "tank@third"');
+    # Mount snapshots
+    assert_script_run('mkdir -p /mnt/tank/{initial,second}');
+    assert_script_run('mount -o ro -t zfs tank@initial /mnt/tank/initial');
+    assert_script_run('zfs clone -o mountpoint=/mnt/tank/second tank@second tank/2nd_second');
+    assert_script_run('stat /mnt/tank/initial/Big_Buck_Bunny_8_seconds_bird_clip.ogv');
+    assert_script_run('cd /mnt/tank/initial/; md5sum -c /var/tmp/Big_Buck_Bunny_8_seconds_bird_clip.ogv.md5sum');
+    assert_script_run('! stat /mnt/tank/initial/test_unzip.zip');
+    assert_script_run('! stat /mnt/tank/second/Big_Buck_Bunny_8_seconds_bird_clip.ogv');
```
[the remark this answers is not in the document]
Comment (grisu48) [reply tests/console/zfs.pm:149]: I don't understand. You mean md5sum without the stat in line 146 is enough or you would like to have md5sum for `test_unzip.zip` as well?

[the remark this answers is not in the document]
Comment (jankohoutek) [reply tests/console/zfs.pm:149]: I mean stat also files on the cloned `/mnt/tank/second/` it 2 different approaches. 
`/mnt/tank/initial` is snapshot mounted ro by `mount.zfs`, which is not but recommended, but supported. 
`/mnt/tank/second/` is clone created from the snapshot and mounted by `zfs mount`.
I would even recommend to also add recommended way of access to the snapshot. Snaps are stored in the `.zfs/snapshot` directory within the root of the snapped file system, in this case it should be in `/tank/.zfs/snapshot/initial`

File: tests/console/zfs.pm (hunk 8 above)
Comment (grisu48) [reply tests/console/zfs.pm:152]: `zfs umount` seems to not work with snapshots: http://phoenix-openqa.qam.suse.de/tests/2721#step/zfs/180

However, I added a `zfs umount` check at the end of the test

File: tests/console/zfs.pm (hunk 9 above)
[the remark this answers is not in the document]
Comment (grisu48) [reply tests/console/zfs.pm:149]: Good point! I'm busy today and then gone for two weeks - if this is not a showstopper I would like to push the PR and apply an update with those points when I'm back.

Would that work for you?

**Comment by grisu48** [comment]
Improvement suggestions will be collected in https://progress.opensuse.org/issues/70981

**Pull request merged**